packages/apollo-engine-reporting/src/extension.ts

```diff
@@ -208,6 +211,177 @@ export class EngineReportingExtension<TContext = any>
+// a trace (in protobuf Trace format) for that single request. When the request
+// is done, it passes the Trace back to its associated EngineReportingAgent via
+// the addTrace callback. This class isn't for direct use; its constructor is a
+// private API for communicating with EngineReportingAgent.
+export const plugin = <TContext>(
+  options: EngineReportingOptions<TContext> = Object.create(null),
+  addTrace: (args: AddTraceArgs) => Promise<void>,
+  // schemaHash: string,
+): ApolloServerPlugin<TContext> => {
+  const logger: Logger = options.logger || console;
+  const generateClientInfo: GenerateClientInfo<TContext> =
+    options.generateClientInfo || defaultGenerateClientInfo;
+
+
+  return {
+    requestDidStart(requestContext) {
+      let queryString: string | undefined;
+      const treeBuilder: EngineReportingTreeBuilder =
+        new EngineReportingTreeBuilder({
+          rewriteError: options.rewriteError,
+          logger: requestContext.logger || logger,
+        });
+
+      const metrics: NonNullable<typeof requestContext['metrics']> =
+        ((requestContext as Mutable<typeof requestContext>)
+          .metrics = requestContext.metrics || Object.create(null));
+
+      treeBuilder.startTiming();
+      metrics.startHrTime = treeBuilder.startHrTime;
+
+      if (requestContext.request.http) {
+        treeBuilder.trace.http = new Trace.HTTP({
+          method:
+            Trace.HTTP.Method[
+              requestContext.request.http
+                .method as keyof typeof Trace.HTTP.Method
+            ] || Trace.HTTP.Method.UNKNOWN,
+          // Host and path are not used anywhere on the backend, so let's not bother
+          // trying to parse request.url to get them, which is a potential
+          // source of bugs because integrations have different behavior here.
+          // On Node's HTTP module, request.url only includes the path
+          // (see https://nodejs.org/api/http.html#http_message_url)
+          // The same is true on Lambda (where we pass event.path)
+          // But on environments like Cloudflare we do get a complete URL.
+          host: null,
+          path: null,
+        });
+      }
+
+      let preflightDone: boolean = false;
+      function ensurePreflight() {
+        if (preflightDone) return;
+        preflightDone = true;
+
+        if (options.sendHeaders) {
+          if (requestContext.request.http && treeBuilder.trace.http) {
+            makeHTTPRequestHeaders(
+              treeBuilder.trace.http,
+              requestContext.request.http.headers,
+              options.sendHeaders,
+            );
+          }
+        }
+
+        if (metrics.persistedQueryHit) {
+          treeBuilder.trace.persistedQueryHit = true;
+        }
+        if (metrics.persistedQueryRegister) {
+          treeBuilder.trace.persistedQueryRegister = true;
+        }
+
+        // Generally, we'll get queryString here and not parsedQuery; we only get
```

Comment:
This comment can go away. You removed the code that handles the `parsedQuery` case which is good because it has been irrelevant for a while (like, since requestPipeline was made I think?)

Also the thing about getting documentAST from executeDidStart isn't true any more either. I can't figure out why we did that instead of just looking on requestContext.document...

Reply:
There was some juggling going on to accommodate eventually-resolved data going on, and at one point an introduction of `didResolveOperation` that could have been related to the once-adopted, questionably-necessary pattern, but I also couldn't concisely figure out all the past reasoning.

I'm going to try to squash some of the naming consistencies on the interface that `addTrace` accepts too (e.g. `document` vs. `documentAST`, `queryString` vs. `source`.

Reply:
Removed comment and `queryString` assignment with fc966a4b5d383d7a0d893839e4b05d9c5b87afaa.

Reply:
And f4aad306 changes the parameter names to `AddTraceArgs` to align with those used in the plugin API and the request pipeline (`documentAST` 👉 `document` and `queryString` 👉 `source`, respectively).